### PR TITLE
[SPARK-34412][SQL] RemoveNoopOperators should only remove trivial projects

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR makes the RemoveNoopOperators only remove attribute only projects.

### Why are the changes needed?
RemoveNoopOperators can remove non-trivial projections that contain actual expressions instead of just attributes, when that expression reuses the expression id of its input. This is a problem with self unions.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit test + regression test.

### CODE REVIEW

**Review by cloud-fan — APPROVED** [review]
LGTM if tests pass

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveNoopOperatorsSuite.scala [hunk 1]
```diff
@@ -44,6 +45,22 @@ class RemoveNoopOperatorsSuite extends PlanTest {
     comparePlans(optimized, testRelation)
   }
 
+  test("Do not remove projections with non-attribute expressions that reuse expr ids") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveNoopOperatorsSuite.scala:48]: It seems that all newly added UTs fails with the following exception.
```
org.apache.spark.sql.catalyst.errors.package$TreeNodeException:
The structural integrity of the input plan is broken in org.apache.spark.sql.catalyst.analysis.SimpleAnalyzer., tree:
```

Comment (hvanhovell) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveNoopOperatorsSuite.scala:48]: Yeah, let me take another look. Internally I am developing against a slightly older version of the codebase, so there might be some differences.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -494,11 +494,14 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
 object RemoveNoopOperators extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
     // Eliminate no-op Projects
-    case p @ Project(_, child) if child.sameOutput(p) => child
+    case p @ Project(projectList, child)
+      if projectList.forall(isAttribute) && child.sameOutput(p) => child
```
Comment (tanelk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:498]: Would this affect other places where the `sameOutput` method is used? It kinda looks like fixing a symptom and the real problem is a bit deeper. The `sameOutput` is used only in few places, but it internally uses `semanticEquals` that is very common.

Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:498]: In the `LogicalPlan` there is this comment:
```
  /**
   * This method checks if the same `ExprId` refers to an unique attribute in a plan tree.
   * Some plan transformers (e.g., `RemoveNoopOperators`) rewrite logical
   * plans based on this assumption.
   */
  def checkIfExprIdsAreGloballyUnique(plan: LogicalPlan): Boolean = {
    checkIfSameExprIdNotReused(plan) && hasUniqueExprIdsForOutput(plan)
  }
```

It sounds like the `RemoveNoopOperators` is correct and the deeper issue is with the `ExprId` getting reused.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @hvanhovell .

**Pull request closed without merging**